MRG: use  dipole output for sparse inverse methods

joewalter opened the pull request:
This PR adds dipole output to our sparse solvers.

Visualization is currently done with `plot_dipole_locations` looping over dipoles or by giving the list of dipoles to `plot_dipole_locations`

agramfort commented on mne/inverse_sparse/mxne_inverse.py:
```diff
@@ -161,11 +162,45 @@ def _make_sparse_stc(X, active_set, forward, tmin, tstep,
+        else:
+            amplitude = linalg.norm(X_, ord=None, axis=0)
+            i_ori = (X_ / amplitude).T
+        dipoles.append(Dipole(times, i_pos, amplitude, i_ori, gof))
```
do we loose the time varying orientation of the dipoles?

joewalter replied:
dipole orientation can vary between time samples, i_ori is activation matrix (3 dipoles per location) normalized to 1 and rotated to xyz coordinate system

agramfort commented on mne/inverse_sparse/mxne_inverse.py:
```diff
@@ -161,11 +162,45 @@ def _make_sparse_stc(X, active_set, forward, tmin, tstep,
+def _make_dipoles(X, active_set, forward, tmin, tstep, M, M_estimated,
+                  verbose=None):
+
+    times = np.arange(tmin, tmin + X.shape[1] * tstep, tstep)
```
I prefer:

times = tmin + tstep * np.arange(X.shape[1])

joewalter replied:
ok, done

agramfort commented:
@yousrabk can you have a look?

joewalter commented:
I renamed mxne_inverse._make_dipoles to mxne_inverse._make_dipoles_sparse to avoid duplication of function names with the function called by rap music

joewalter commented:
Basically I am applying our sparse imaging methods with different setting and then testing that the location of the first active source in the stc is the same as the dipole position. To speed up, we could  test only 1 setting per method. The last commit also contains a small change to avoid a divide by 0 warning in the computation of the GOF, which only appears for TF-MxNE due to the applied Hann window.

agramfort commented:
ok

see how much you can reduce test time while still testing what needs to be

thx

joewalter commented:
Can someone give me a hint how to get rid of this CircleCI error?

agramfort commented:
I restarted appveyor and circleci

joewalter commented:
@agramfort the same errors are still present in appveyor and circleci.

agramfort commented:
running gamma map example i see this scrambled fig:

<img width="631" alt="screen shot 2017-07-08 at 14 01 58" src="https://user-images.githubusercontent.com/161052/27985143-12f967f4-63e6-11e7-90df-64aeeed7d773.png">

regarding failures on circle and appveyor maybe try to plot less things. Plotting all dipoles etc can maybe use a lot of RAM.

btw we should have a way to plot the dipoles with a glass brain using nilearn

https://nilearn.github.io/auto_examples/01_plotting/plot_demo_glass_brain.html#sphx-glr-auto-examples-01-plotting-plot-demo-glass-brain-py

and overlay the dipoles like BESA. See upper right plot at:

http://wiki.besa.de/index.php?title=File:DipoleFromSource2arrow.jpg

here we ave 4 dipoles for gamma map and we end up with 4 figures and there is no way to have position and orientation of the 4 dipoles in 1 figure.

@jaeilepp feeling bored?

joewalter commented:
What I have locally is a visualization showing dipole positions and orientations of all dipoles in orthoview. Basically I check the dipole times of the dipoles (after concatenation to one big dipole) and then plot all dipoles with orientation which are active at the same time point. But a glass brain might be better, that's why I loop over dipoles in the example and show separate figures. 

Ideally, for spatio-temporal dipoles, we would have an interactive dipole plotting tool showing dipole activations and locations in the glass brain and which allows one to select a specific dipole and time point in the activation and then shows the orthoview image for this dipole.

joewalter commented:
Ok, I will try to reduce the number of plots in the examples to see if this fixed the errors

jaeilepp commented:
I tried debugging the circle failure, but was unable to reproduce it on SSH. The failure is caused by this line: ``tail $CIRCLE_ARTIFACTS/log.txt | grep "auto_tutorials\[100%\] plot_"``. In fact, I don't even understand the purpose of this line

> btw we should have a way to plot the dipoles with a glass brain using nilearn

Can I copy the drawing of glass brain from nilearn to make this happen?

jaeilepp commented:
It seems the circle error is this:
```
WARNING: /home/ubuntu/mne-python/examples/inverse/plot_time_frequency_mixed_norm_inverse.py failed to execute correctly:Traceback (most recent call last):
  File "/home/ubuntu/mne-python/sphinx-gallery/sphinx_gallery/gen_rst.py", line 454, in execute_code_block
    exec(code_block, example_globals)
  File "<string>", line 9, in <module>
AttributeError: 'list' object has no attribute 'crop'
```

joewalter commented:
Thanks @jaeilepp! Found the issue. Working on the fix now.

joewalter commented:
@agramfort I cannot reproduce the scrambled image in the gamma_map example locally

joewalter commented:
All checks pass now. Setting to merge.
@agramfort, @jaeilepp can you please double check the rotation of the coordinate systems in `_make_dipoles_sparse` (line 194-197 in `mxne_inverse`).

I simplified the examples and removed some of the dipole plots (leaving the option to plot all dipoles as a comment). The examples now compute the inverse solution only once (returning dipoles) and generate the stc with `make_sparse_stc_from_dipoles`.

@jaeilepp are you working on a separate PR for the glass brain plotting?

agramfort commented:
> Can I copy the drawing of glass brain from nilearn to make this happen?

yes but we already rely on nilearn for plotting so maybe add a weak
dependency on nilearn for this plot

I'll look at the details when I have time

jaeilepp commented:
The problem with adding a dependency is that AFAIK the nilearn plotting function wants a filename. So we would have to save the data first.

> @jaeilepp are you working on a separate PR for the glass brain plotting?

Yeah I'll do it after this gets merged, so I can use the functions in this for creating the source estimates.

Eric89GXL commented:
> The problem with adding a dependency is that AFAIK the nilearn plotting function wants a filename. So we would have to save the data first.

Can you look into making a PR to `nilearn` to allow it to take an appropriate instance (of a `nibabel` object I guess?) instead?

Eric89GXL commented on doc/whats_new.rst:
```diff
@@ -58,6 +58,10 @@ Changelog
 
     - Add :func:`mne.inverse_sparse.mxne_optim.dgap_l21l1` for computing the duality gap for TF-MxNE as the new stopping criterion by `Daniel Strohmeier`_
+
+    - Add option to return dipole objects in sparse source imaging methods by `Daniel Strohmeier`_
```
```
:class:`Dipole` objects
```

Eric89GXL commented on examples/inverse/plot_time_frequency_mixed_norm_inverse.py:
```diff
@@ -36,11 +36,14 @@
-from mne.viz import plot_sparse_source_estimates
+from mne.inverse_sparse import tf_mixed_norm, make_sparse_stc_from_dipoles
+from mne.viz import (plot_sparse_source_estimates,
+                     plot_dipole_locations, plot_dipole_amplitudes)
```
The References in this example are not properly formatted

joewalter commented:
@agramfort do you have time to have a look? I'd like to use this PR here as the basis for reopening #3641

agramfort commented on examples/inverse/plot_gamma_map_inverse.py:
```diff
@@ -64,3 +77,18 @@
+
+###############################################################################
+# Generate stc from dipoles
+stc = make_sparse_stc_from_dipoles(dipoles, forward)
```
can you clarify why you need to pass forward and not just the src?

joewalter replied:
I'm also using `forward['source_rr']` to find the index of the source position which corresponds to the dipole location.  `rr` is also in `src`, but for each source space separately, right? So I'd need to build the matrix first by looping of src. Is that better?

agramfort replied:
if you can make it work with src yes I think it's cleaner.

agramfort commented on examples/inverse/plot_time_frequency_mixed_norm_inverse.py:
```diff
@@ -82,18 +85,36 @@
+plot_dipole_amplitudes(dipoles)
+
+# Plot dipole location of the strongest dipole with MRI slices
+idx = np.argmax([np.amax(np.abs(dip.amplitude)) for dip in dipoles])
```
you can use np.max (not amax)

agramfort commented on examples/inverse/plot_mixed_norm_inverse.py:
```diff
@@ -63,14 +64,42 @@
+plot_dipole_amplitudes(dipoles)
+
+# Plot dipole location of the strongest dipole with MRI slices
+idx = np.argmax([np.amax(np.abs(dip.amplitude)) for dip in dipoles])
```
you can use np.max (not amax)

agramfort commented on mne/inverse_sparse/mxne_inverse.py:
```diff
@@ -161,11 +162,104 @@ def _make_sparse_stc(X, active_set, forward, tmin, tstep,
+    M_norm = np.sqrt(np.sum(M ** 2, axis=0))
+    gof[M_norm > 0.0] = (np.sqrt(np.sum(M_est ** 2., axis=0))[M_norm > 0.0] /
+                         M_norm[M_norm > 0.0] * 100.)
+
```
I am bit concerned that GOF for each dipole is not meaningful with such multi dipole fit approach?

wdyt of having a GOF which is dipole specific ?

basically to store the variance each dipole can explain individually

joewalter replied:
Ok this wasn't the GOF, thanks for the hint. Corrected it with my last commit. 

I think it is misleading to compute values for each dipole individually as we fit a multidipole model. So the GOF should also be computed for the multidipole model. Especially as the sum over individual GOFs must not be equivalent to the GOF of the full model. Iirc the same is done in the RAP-Music code

agramfort commented on mne/viz/misc.py:
```diff
@@ -586,6 +586,8 @@ def plot_dipole_amplitudes(dipoles, colors=None, show=True):
     show : bool
         Show figure if True.
+    linewidth : int
+        Line width in 2D plot.
```
please revert this. You can fix this after plotting by working with the outputed fig object.

joewalter replied:
I found it strange that the two plots showing dipole activations (plot_dipole_amplitudes and plot_sparse_source_estimates) have different linewidth

agramfort commented:
that's it for me

agramfort commented:
ok fair enough. Yes it's done also indeed in rap-music

joewalter commented:
@agramfort do you still have these scrambled figs on this branch?

Can someone restart Travis? There's a download error on one Travis run. The other runs are green.

agramfort commented:
thx heaps @joewalter  !

agramfort commented:
I just had to push a cosmit:

https://github.com/mne-tools/mne-python/commit/9d9fbd30da64c0abaf574ea8072ddf579791584a
